app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/AddPictureActivity.kt

```diff
@@ -0,0 +1,126 @@
+                        Manifest.permission.CAMERA
+                ) == PackageManager.PERMISSION_GRANTED -> {
+                    setupCameraPreviewView()
+                }
+                shouldShowRequestPermissionRationale(Manifest.permission.CAMERA) -> {
+                    permissionNeededDialog()
+                }
+                else -> {
+                    requestPermissionLauncher.launch(Manifest.permission.CAMERA)
+                }
+            }
+        } else {
+            permissionNeededDialog()
+        }
+    }
+
+    private fun cameraIsAvailable(): Boolean {
+        return applicationContext.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA)
+    }
+
+    private fun setupRequestPermissionLauncher() {
+        requestPermissionLauncher =
+                registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
+                    if (isGranted) {
+                        setupCameraPreviewView()
+                        Toast.makeText(applicationContext, getString(R.string.AddPicture_permissionGrantedToast), Toast.LENGTH_SHORT).show()
+                    } else {
+                        permissionNeededDialog()
+                    }
+                }
+    }
+
+    /*
+    Needs to be set on a button
+     */
+    private fun setupOpenLabelSelectionLayoutListener() {
+        openLabelSelectionLayoutListener = View.OnClickListener { it: View ->
+            val builder = AlertDialog.Builder(this)
+
+            builder.apply {
+                setTitle(getString(R.string.AddPicture_labelSelectionDialogTitle))
+                setItems(dummyLabels) { dialog, label_index ->
+                    (it as Button).text = dummyLabels[label_index]
+                    it.background.setTint(getColor(R.color.button_set))
+                    it.setTextColor(getColor(R.color.black))
+                }
+            }
+            val dialog = builder.create()
+            dialog.show()
+        }
+    }
+
+    private fun setupCameraPreviewView() {
+        setContentView(R.layout.activity_camera)
+        findViewById<Button>(R.id.selectLabelButton).setOnClickListener(openLabelSelectionLayoutListener)
+        cameraProviderFuture = ProcessCameraProvider.getInstance(this)
+        cameraProviderFuture.addListener(Runnable {
+            val cameraProvider = cameraProviderFuture.get()
+            bindPreview(cameraProvider)
+        }, ContextCompat.getMainExecutor(this))
+    }
+
+    private fun bindPreview(cameraProvider: ProcessCameraProvider) {
+        val preview: Preview = Preview.Builder().build()
+        val cameraSelector: CameraSelector =
+                CameraSelector.Builder().requireLensFacing(CameraSelector.LENS_FACING_BACK).build()
+
+        preview.setSurfaceProvider(findViewById<PreviewView>(R.id.cameraPreviewView).surfaceProvider)
+
+        val camera = cameraProvider.bindToLifecycle(this as LifecycleOwner, cameraSelector, preview)
```

Comment:
```suggestion
        val camera = cameraProvider.bindToLifecycle(this, cameraSelector, preview)
```
This upcast is redundant.

Reply:
Yes, fixed